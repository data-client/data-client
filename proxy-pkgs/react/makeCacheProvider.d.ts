export * from '@data-client/react/makeCacheProvider';
