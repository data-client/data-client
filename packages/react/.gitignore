/lib
/dist
/legacy
/index.d.ts
/next.d.ts
