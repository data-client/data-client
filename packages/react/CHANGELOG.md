# Change Log

## 7.3.1

### Patch Changes

- 9157e9eba8: Fix: SSR previously would never unsuspend cache provider
  (NetworkManager.allSettled() must return undefined if nothing in flight)
- Updated dependencies [9157e9eba8]
  - @rest-hooks/core@4.3.1

## 7.3.0

### Minor Changes

- 775352c9ab: useController().fetch resolves to denormalized form
- 775352c9ab: Add /next export for early adoption of breaking changes

### Patch Changes

- 54aad5e028: FIx ResultEntry with schema
- Updated dependencies [54aad5e028]
- Updated dependencies [775352c9ab]
- Updated dependencies [775352c9ab]
  - @rest-hooks/core@4.3.0

## 7.2.12

### Patch Changes

- dbc4d0b843: Polling intervals double check cancellation before dispatching fetch
- Updated dependencies [dbc4d0b843]
- Updated dependencies [dbc4d0b843]
  - @rest-hooks/core@4.2.11

## 7.2.11

### Patch Changes

- f927a15c14: Add ErrorTypes export
- Updated dependencies [f927a15c14]
  - @rest-hooks/core@4.2.10

## 7.2.10

### Patch Changes

- ebe0fcc15c: fix: PnP compatibility

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

### [7.2.9](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.8...@rest-hooks/react@7.2.9) (2023-03-21)

**Note:** Version bump only for package @rest-hooks/react

### [7.2.8](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.7...@rest-hooks/react@7.2.8) (2023-03-19)

### 🐛 Bug Fix

- Controller type generics refer to findable types ([449d326](https://github.com/data-client/rest-hooks/commit/449d326b0163cd3f915dedf7406c6a18857c1550))

### [7.2.7](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.6...@rest-hooks/react@7.2.7) (2023-03-15)

### 💅 Enhancement

- Mark useController() as 'use client' since it has context ([5f25fba](https://github.com/data-client/rest-hooks/commit/5f25fbadd4747cd47962b27ba1400acba4753310))

### 📦 Package

- Update babel packages ([#2487](https://github.com/data-client/rest-hooks/issues/2487)) ([3fc9efc](https://github.com/data-client/rest-hooks/commit/3fc9efc0bfc818ae7b4a1113cf6f7daa8bbcba8c))

### [7.2.6](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.5...@rest-hooks/react@7.2.6) (2023-03-12)

### 📝 Documentation

- Update links in readme ([1141c0b](https://github.com/data-client/rest-hooks/commit/1141c0b41c05250e5455457d00a2c54b631e4f91))

### [7.2.5](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.4...@rest-hooks/react@7.2.5) (2023-03-02)

### 🚀 Features

- Add mixin schema.Entity() ([#2449](https://github.com/data-client/rest-hooks/issues/2449)) ([6c5d5db](https://github.com/data-client/rest-hooks/commit/6c5d5db7f5856c21927e9fa389c09ead11e93ea0))

### 💅 Enhancement

- Allow 'false' as Endpoint.sideEffect ([#2452](https://github.com/data-client/rest-hooks/issues/2452)) ([69b6a04](https://github.com/data-client/rest-hooks/commit/69b6a049861eeefc608f5a5df86b9833021961df))

### 📦 Package

- Update babel packages ([#2457](https://github.com/data-client/rest-hooks/issues/2457)) ([45c4a0a](https://github.com/data-client/rest-hooks/commit/45c4a0ab4ebd6112db75e8c6f09e5ad1add74c8b))
- Update JS test packages ([#2456](https://github.com/data-client/rest-hooks/issues/2456)) ([6732e63](https://github.com/data-client/rest-hooks/commit/6732e63a10ab3cebe70f22e44429c5edae65186e))

### [7.2.4](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.3...@rest-hooks/react@7.2.4) (2023-02-21)

### 💅 Enhancement

- Use null-object over Object.hasOwn when possible ([#2441](https://github.com/data-client/rest-hooks/issues/2441)) ([28f3fe8](https://github.com/data-client/rest-hooks/commit/28f3fe8890da119c7fa6498d4617a66da74ed0b4))

### [7.2.3](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.2...@rest-hooks/react@7.2.3) (2023-02-14)

**Note:** Version bump only for package @rest-hooks/react

### [7.2.2](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.1...@rest-hooks/react@7.2.2) (2023-02-11)

**Note:** Version bump only for package @rest-hooks/react

### [7.2.1](https://github.com/data-client/rest-hooks/compare/@rest-hooks/react@7.2.0...@rest-hooks/react@7.2.1) (2023-01-30)

### 📝 Documentation

- Update github organization to data-client ([#2396](https://github.com/data-client/rest-hooks/issues/2396)) ([45faab1](https://github.com/data-client/rest-hooks/commit/45faab1962cad292d1f77a0a997e8c321a6917db))

## [7.2.0](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.7...@rest-hooks/react@7.2.0) (2023-01-29)

### 🚀 Features

- Add controller.invalidateAll() ([#2382](https://github.com/coinbase/rest-hooks/issues/2382)) ([d145057](https://github.com/coinbase/rest-hooks/commit/d145057bbafb4a2efee5a31fdf2e32f1b16971b6))
- Add Interceptors ([#2380](https://github.com/coinbase/rest-hooks/issues/2380)) ([53653ba](https://github.com/coinbase/rest-hooks/commit/53653ba9462510938f57cfe4ff907fda2efcc874))

### 💅 Enhancement

- Improve strictNullChecks: false compatibility ([#2386](https://github.com/coinbase/rest-hooks/issues/2386)) ([1cfb5a5](https://github.com/coinbase/rest-hooks/commit/1cfb5a53564719952ee30b1b3160cd32f6bd2603))

### 🐛 Bug Fix

- Conditional null args type handling ([#2386](https://github.com/coinbase/rest-hooks/issues/2386)) ([6d6e0f7](https://github.com/coinbase/rest-hooks/commit/6d6e0f74399452fcf3e78561a69edcb6e8e78a9d))

### [7.1.7](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.6...@rest-hooks/react@7.1.7) (2023-01-23)

### 🐛 Bug Fix

- Add back DenormalizeCacheContext for older versions of redux ([8e0acab](https://github.com/coinbase/rest-hooks/commit/8e0acabbd86ca8cd2cd45d07c8e815df8c7375d9))

### [7.1.6](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.5...@rest-hooks/react@7.1.6) (2023-01-23)

### 🚀 Features

- makeRenderRestHook(CacheProvider) and React 18 testing ([#2328](https://github.com/coinbase/rest-hooks/issues/2328)) ([0e9e51e](https://github.com/coinbase/rest-hooks/commit/0e9e51e3bce3c9c978888a734c43be8d1fe3ae55))

### 💅 Enhancement

- Stop uncaught promises ([aac6fe9](https://github.com/coinbase/rest-hooks/commit/aac6fe9523c5eeee0068a227bcf39798c80fc55f))

### 📦 Package

- Update linting packages ([#2376](https://github.com/coinbase/rest-hooks/issues/2376)) ([305debd](https://github.com/coinbase/rest-hooks/commit/305debd269216444519d80c351fd63805a974343))

### 📝 Documentation

- Add atomic mutations concept ([#2369](https://github.com/coinbase/rest-hooks/issues/2369)) ([a56380b](https://github.com/coinbase/rest-hooks/commit/a56380b8cb39348572b99013945fbb217566662c))
- Update resource def best practices for github example ([#2363](https://github.com/coinbase/rest-hooks/issues/2363)) ([3144584](https://github.com/coinbase/rest-hooks/commit/3144584bb773c7c44362f08adedd89107d3b2313))

### [7.1.5](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.4...@rest-hooks/react@7.1.5) (2023-01-14)

### 🐛 Bug Fix

- Compatibility with React 18 StrictMode ([e7174a3](https://github.com/coinbase/rest-hooks/commit/e7174a3c2aac3fd611d2e1305a4fe4927ef50e38))

### 📦 Package

- Update linting packages ([#2351](https://github.com/coinbase/rest-hooks/issues/2351)) ([d3498ea](https://github.com/coinbase/rest-hooks/commit/d3498ea396dfbfdc745ec6e68920a714d8870fe8))

### [7.1.4](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.3...@rest-hooks/react@7.1.4) (2023-01-12)

### 📦 Package

- Update babel packages ([#2348](https://github.com/coinbase/rest-hooks/issues/2348)) ([5d0a68e](https://github.com/coinbase/rest-hooks/commit/5d0a68ea00b021effeae185fcdea415cf50c0328))

### 📝 Documentation

- Add TypeScriptEditor to aid type demos ([#2355](https://github.com/coinbase/rest-hooks/issues/2355)) ([2fb44a9](https://github.com/coinbase/rest-hooks/commit/2fb44a971d80a3a170d37dbc7ecf459c7da39722))

### [7.1.3](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.2...@rest-hooks/react@7.1.3) (2022-12-23)

### 🚀 Features

- Add ctrl.setResponse, ctrl.setError ([#2331](https://github.com/coinbase/rest-hooks/issues/2331)) ([46e3b63](https://github.com/coinbase/rest-hooks/commit/46e3b630dc66af54979eafbfc3c49847a4553f3f))

### 📦 Package

- Update babel packages ([#2339](https://github.com/coinbase/rest-hooks/issues/2339)) ([2cc99b9](https://github.com/coinbase/rest-hooks/commit/2cc99b99aeece58b0e7674ca80d3372555612c63))

### 📝 Documentation

- Add 7, 7.1 blog; new intro ([#2326](https://github.com/coinbase/rest-hooks/issues/2326)) ([b76c134](https://github.com/coinbase/rest-hooks/commit/b76c134cd8d0675c0eef7413ba3fddb6262443cb))
- Updates to the docs ([#2335](https://github.com/coinbase/rest-hooks/issues/2335)) ([73c8d9f](https://github.com/coinbase/rest-hooks/commit/73c8d9f4b403412cd766955305461aa7d8ebb462))
- Yet another docs change pr ([#2338](https://github.com/coinbase/rest-hooks/issues/2338)) ([52c2f83](https://github.com/coinbase/rest-hooks/commit/52c2f83d20106f8557ef8538a348c4bbac77dd53))

### [7.1.2](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.1...@rest-hooks/react@7.1.2) (2022-12-06)

### 💅 Enhancement

- Reduce bundlesize via custom Object.hasOwn polyfill ([#2322](https://github.com/coinbase/rest-hooks/issues/2322)) ([1f8009d](https://github.com/coinbase/rest-hooks/commit/1f8009d9849cfb784f072a1078b4d12b37d59a04))

### 📝 Documentation

- Random improvements and robots.txt again ([#2318](https://github.com/coinbase/rest-hooks/issues/2318)) ([b146f82](https://github.com/coinbase/rest-hooks/commit/b146f821f4cef623d410692449555190deb9a9e0))

### [7.1.1](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.1.0...@rest-hooks/react@7.1.1) (2022-12-05)

### 💅 Enhancement

- Block dispatches after unmount ([#2307](https://github.com/coinbase/rest-hooks/issues/2307)) ([55af3e6](https://github.com/coinbase/rest-hooks/commit/55af3e693c310f81511b254c1c2451d3bcbd7342))
- Include Object.hasOwn polyfill ([#2309](https://github.com/coinbase/rest-hooks/issues/2309)) ([14b93f6](https://github.com/coinbase/rest-hooks/commit/14b93f67f0589df5813909e0c1acd4cacad0a3ee))

### 📦 Package

- Update babel packages ([#2308](https://github.com/coinbase/rest-hooks/issues/2308)) ([e3ee5ee](https://github.com/coinbase/rest-hooks/commit/e3ee5ee57431971ba4bdb47b48ed89933412374c))

### 📝 Documentation

- Lots of fixes ([#2295](https://github.com/coinbase/rest-hooks/issues/2295)) ([2d151e8](https://github.com/coinbase/rest-hooks/commit/2d151e824bac674f35b20f684defffd26c1409a1))

## [7.1.0](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@7.0.0...@rest-hooks/react@7.1.0) (2022-11-22)

### 🚀 Features

- Add LogoutManager ([e9e10a8](https://github.com/coinbase/rest-hooks/commit/e9e10a88df2bbc36bfaf3498c00cc35b657ada8f))
- Add useLive() ([#2287](https://github.com/coinbase/rest-hooks/issues/2287)) ([e1623f9](https://github.com/coinbase/rest-hooks/commit/e1623f9868a6ab7e7799ef386a087468f8b6e006))
- Make Controller the middleware API ([#2290](https://github.com/coinbase/rest-hooks/issues/2290)) ([0f0302d](https://github.com/coinbase/rest-hooks/commit/0f0302d6e95faea67cc3283bc013de06b9dcc840))
- **react-native:** Sub/unsub based on react-navigation focus ([#2291](https://github.com/coinbase/rest-hooks/issues/2291)) ([efc926f](https://github.com/coinbase/rest-hooks/commit/efc926f0a1036d98063c80b98d24338209f21c18))

### 📝 Documentation

- Add nextjs demo ([b650618](https://github.com/coinbase/rest-hooks/commit/b6506180ef41a73eb4c926eef3786e0394c3a2c3))
- Add useLive() docs ([3917c1c](https://github.com/coinbase/rest-hooks/commit/3917c1c3490794d0a7a987c7d6b3a255b25943ee))

## [7.0.0](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@6.0.0...@rest-hooks/react@7.0.0) (2022-11-16)

### ⚠ 💥 BREAKING CHANGES

- Hooks may fetch more than previously

### 🚀 Features

- Revalidate on react-navigation focus for useFetch() ([6eba85a](https://github.com/coinbase/rest-hooks/commit/6eba85a80aef919b066650e3a7f6ff38af831f8b))

### [0.2.1](https://github.com/coinbase/rest-hooks/compare/@rest-hooks/react@0.2.0...@rest-hooks/react@0.2.1) (2022-11-13)

### 📝 Documentation

- Update badge links ([e79c685](https://github.com/coinbase/rest-hooks/commit/e79c6853e9414127c6eeaee784dc3f33546b9630))

## 0.2.0 (2022-11-13)

### 🚀 Features

- Add packages react and redux ([#2270](https://github.com/coinbase/rest-hooks/issues/2270)) ([6e28f5f](https://github.com/coinbase/rest-hooks/commit/6e28f5f465b6f4f5d444b56234f212863aeade31))
