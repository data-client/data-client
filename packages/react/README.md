# [![Reactive Data Client](./data_client_logo_and_text.svg?sanitize=true)](https://dataclient.io)

[![CircleCI](https://circleci.com/gh/data-client/data-client/tree/master.svg?style=shield)](https://circleci.com/gh/data-client/data-client)
[![Coverage Status](https://img.shields.io/codecov/c/gh/data-client/data-client/master.svg?style=flat-square)](https://app.codecov.io/gh/data-client/data-client?branch=master)
[![Percentage of issues still open](https://isitmaintained.com/badge/open/data-client/data-client.svg)](https://github.com/data-client/data-client/issues "Percentage of issues still open")
[![bundle size](https://img.shields.io/bundlephobia/minzip/@data-client/react?style=flat-square)](https://bundlephobia.com/result?p=@data-client/react)
[![npm version](https://img.shields.io/npm/v/@data-client/react.svg?style=flat-square)](https://www.npmjs.com/package/@data-client/react)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![Chat](https://img.shields.io/discord/768254430381735967.svg?style=flat-square&colorB=758ED3)](https://discord.gg/35nb8Mz)

Define your [async methods](https://dataclient.io/docs/getting-started/resource). Use them [synchronously in React](https://dataclient.io/docs/getting-started/data-dependency). [Instantly mutate](https://dataclient.io/docs/getting-started/mutations) the data and automatically update all usages.

For [REST](https://dataclient.io/rest), [GraphQL](https://dataclient.io/graphql), [Websockets+SSE](https://dataclient.io/docs/api/Manager#data-stream) and [more](https://dataclient.io/docs/guides/img-media)

<div align="center">

**[📖Read The Docs](https://dataclient.io/docs)** &nbsp;|&nbsp; [🏁Getting Started](https://dataclient.io/docs/getting-started/installation) &nbsp;|&nbsp;
[🎮Todo Demo](https://stackblitz.com/github/data-client/data-client/tree/master/examples/todo-app?file=src%2Fpages%2FHome%2FTodoList.tsx) &nbsp;|&nbsp;
[🎮Github Demo](https://stackblitz.com/github/data-client/data-client/tree/master/examples/github-app?file=src%2Fpages%2FIssueList.tsx) &nbsp;|&nbsp;
[🎮NextJS SSR Demo](https://stackblitz.com/github/data-client/data-client/tree/master/examples/nextjs?file=pages%2FAssetPrice.tsx)

</div>

## Installation

```bash
npm install --save @data-client/react @data-client/rest @data-client/test
```

For more details, see [the Installation docs page](https://dataclient.io/docs/getting-started/installation).

## Usage

### Simple [TypeScript definition](https://dataclient.io/rest/api/Entity)

```typescript
class User extends Entity {
  id = '';
  username = '';

  pk() {
    return this.id;
  }
}

class Article extends Entity {
  id = '';
  title = '';
  body = '';
  author = User.fromJS();
  createdAt = new Date(0);

  pk() {
    return this.id;
  }

  static schema = {
    author: User,
    createdAt: Date,
  };
}
```

### Create [collection of API Endpoints](https://dataclient.io/docs/getting-started/resource)

```typescript
const UserResource = createResource({
  path: '/users/:id',
  schema: User,
  optimistic: true,
});

const ArticleResource = createResource({
  path: '/articles/:id',
  schema: Article,
  searchParams: {} as { author?: string },
  optimistic: true,
  paginationField: 'cursor',
});
```

### One line [data binding](https://dataclient.io/docs/getting-started/data-dependency)

```tsx
const article = useSuspense(ArticleResource.get, { id });
return (
  <article>
    <h2>
      {article.title} by {article.author.username}
    </h2>
    <p>{article.body}</p>
  </article>
);
```

### [Reactive Mutations](https://dataclient.io/docs/getting-started/mutations)

```tsx
const ctrl = useController();
return (
  <CreateProfileForm
    onSubmit={data => ctrl.fetch(UserResource.getList.push, { id }, data)}
  />
  <ProfileForm
    onSubmit={data => ctrl.fetch(UserResource.update, { id }, data)}
  />
  <button onClick={() => ctrl.fetch(UserResource.delete, { id })}>Delete</button>
);
```

### [Subscriptions](https://dataclient.io/docs/api/useLive)

```tsx
const price = useLive(PriceResource.get, { symbol });
return price.value;
```

### [Type-safe Imperative Actions](https://dataclient.io/docs/api/Controller)

```tsx
const ctrl = useController();
ctrl.expireAll(ArticleResource.getList);
ctrl.invalidate(ArticleResource.get, { id });
ctrl.invalidateAll(ArticleResource.getList);
ctrl.setResponse(ArticleResource.get, { id }, articleData);
ctrl.fetch(ArticleResource.get, { id });
```

### [Programmatic queries](https://dataclient.io/rest/api/Query)

```tsx
const queryTotalVotes = new Query(
  new schema.All(Post),
  (posts, { userId } = {}) => {
    if (userId !== undefined)
      posts = posts.filter(post => post.userId === userId);
    return posts.reduce((total, post) => total + post.votes, 0);
  },
);

const totalVotes = useCache(queryTotalVotes);
const totalVotesForUser = useCache(queryTotalVotes, { userId });
```

### [Powerful Middlewares](https://dataclient.io/docs/concepts/managers)

```ts
class LoggingManager implements Manager {
  getMiddleware = (): Middleware => controller => next => async action => {
    console.log('before', action, controller.getState());
    await next(action);
    console.log('after', action, controller.getState());
  };

  cleanup() {}
}
```

### [Integrated data mocking](https://dataclient.io/docs/api/Fixtures)

```tsx
const fixtures = [
  {
    endpoint: ArticleResource.getList,
    args: [{ maxResults: 10 }] as const,
    response: [
      {
        id: '5',
        title: 'first post',
        body: 'have a merry christmas',
        author: { id: '10', username: 'bob' },
        createdAt: new Date(0).toISOString(),
      },
      {
        id: '532',
        title: 'second post',
        body: 'never again',
        author: { id: '10', username: 'bob' },
        createdAt: new Date(0).toISOString(),
      },
    ],
  },
  {
    endpoint: ArticleResource.update,
    response: ({ id }, body) => ({
      ...body,
      id,
    }),
  },
];

const Story = () => (
  <MockResolver fixtures={options[result]}>
    <ArticleList maxResults={10} />
  </MockResolver>
);
```

### ...all typed ...fast ...and consistent

For the small price of 9kb gziped. &nbsp;&nbsp; [🏁Get started now](https://dataclient.io/docs/getting-started/installation)

## Features

- [x] ![TS](./typescript.svg?sanitize=true) Strong [Typescript](https://www.typescriptlang.org/) inference
- [x] 🛌 React [Suspense](https://dataclient.io/docs/getting-started/data-dependency#boundaries) support
- [x] 🧵 React 18 [Concurrent mode](https://dataclient.io/docs/guides/render-as-you-fetch) compatible
- [x] 💦 [Partial Hydration Server Side Rendering](https://dataclient.io/docs/guides/ssr)
- [x] 🎣 [Declarative API](https://dataclient.io/docs/getting-started/data-dependency)
- [x] 📝 Composition over configuration
- [x] 💰 [Normalized](https://dataclient.io/docs/concepts/normalization) caching
- [x] 💥 Tiny bundle footprint
- [x] 🛑 Automatic overfetching elimination
- [x] ✨ [Optimistic updates](https://dataclient.io/rest/guides/optimistic-updates)
- [x] 🧘 [Flexible](https://dataclient.io/docs/getting-started/resource) to fit any API design (one size fits all)
- [x] 🔧 [Debugging and inspection](https://dataclient.io/docs/guides/debugging) via browser extension
- [x] 🌳 Tree-shakable (only use what you need)
- [x] 🔁 [Subscriptions](https://dataclient.io/docs/api/useSubscription)
- [x] ♻️ Optional [redux integration](https://dataclient.io/docs/guides/redux)
- [x] 📙 [Storybook mocking](https://dataclient.io/docs/guides/storybook)
- [x] 📱 [React Native](https://facebook.github.io/react-native/) support
- [x] ⚛️ [NextJS](https://dataclient.io/docs/guides/ssr#nextjs) support
- [x] 🚯 [Declarative cache lifetime policy](https://dataclient.io/docs/concepts/expiry-policy)
- [x] 🧅 [Composable middlewares](https://dataclient.io/docs/api/Manager)
- [x] 💽 Global data consistency guarantees
- [x] 🏇 Automatic race condition elimination
- [x] 👯 Global referential equality guarantees

## Examples

- Todo: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/data-client/data-client/tree/master/examples/todo-app) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/data-client/data-client/tree/master/examples/todo-app?file=src%2Fpages%2FHome%2FTodoList.tsx)
- Github: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/data-client/data-client/tree/master/examples/github-app) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/data-client/data-client/tree/master/examples/github-app?file=src%2Fpages%2FIssueList.tsx)
- NextJS: [![GitHub](https://badgen.net/badge/icon/github?icon=github&label)](https://github.com/data-client/data-client/tree/master/examples/nextjs) | [![Sandbox](https://developer.stackblitz.com/img/open_in_stackblitz_small.svg)](https://stackblitz.com/github/data-client/data-client/tree/master/examples/nextjs?file=pages%2FAssetPrice.tsx)
