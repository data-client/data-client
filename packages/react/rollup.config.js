import babel from 'rollup-plugin-babel';
import commonjs from 'rollup-plugin-commonjs';
import filesize from 'rollup-plugin-filesize';
import json from 'rollup-plugin-json';
import resolve from 'rollup-plugin-node-resolve';
import replace from 'rollup-plugin-replace';
import { terser } from 'rollup-plugin-terser';

import pkg from './package.json';
import { typeConfig, typeConfigNext } from '../../config/rollup-utils';

const dependencies = Object.keys(pkg.dependencies)
  .concat(Object.keys(pkg.peerDependencies))
  .filter(
    dep =>
      !['@rest-hooks/use-enhanced-reducer', '@babel/runtime'].includes(dep),
  );

const extensions = ['.js', '.ts', '.tsx', '.mjs', '.json', '.node', '.jsx'];
const nativeExtensions = ['.native.ts', ...extensions];
process.env.NODE_ENV = 'production';

function isExternal(id) {
  return dependencies.some(dep => dep === id || id.startsWith(dep));
}

const configs = [];
if (process.env.BROWSERSLIST_ENV !== 'node12') {
  // browser-friendly UMD build
  configs.push({
    input: 'lib/index.js',
    external: isExternal,
    output: [{ file: pkg.unpkg, format: 'umd', name: 'restHookCore' }],
    plugins: [
      babel({
        exclude: ['node_modules/**', '/**__tests__/**'],
        extensions,
        rootMode: 'upward',
        runtimeHelpers: true,
      }),
      replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
      resolve({ extensions }),
      commonjs({ extensions }),
      json(),
      terser({}),
      filesize({ showBrotliSize: true }),
    ],
  });
  configs.push(typeConfig);
  configs.push(typeConfigNext);
} else {
  // node-friendly commonjs build
  [
    { input: 'lib/index.js', output: pkg.main },
    { input: 'lib/next/index.js', output: 'dist/next.js' },
    { input: 'lib/makeCacheProvider.js', output: 'dist/makeCacheProvider.js' },
  ].forEach(({ input, output }) => {
    configs.push({
      input,
      external: isExternal,
      output: [{ file: output, format: 'cjs' }],
      plugins: [
        babel({
          exclude: ['node_modules/**', '**/__tests__/**', '**/*.d.ts'],
          extensions: nativeExtensions,
          rootMode: 'upward',
          runtimeHelpers: true,
        }),
        replace({ 'process.env.CJS': 'true' }),
        resolve({ extensions: nativeExtensions }),
        commonjs({ extensions: nativeExtensions }),
      ],
    });
  });
}
export default configs;
