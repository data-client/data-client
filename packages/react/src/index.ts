Object.hasOwn =
  Object.hasOwn ||
  /* istanbul ignore next */ function hasOwn(it, key) {
    return Object.prototype.hasOwnProperty.call(it, key);
  };
export {
  PollingSubscription,
  DevToolsManager,
  SubscriptionManager,
  DefaultConnectionListener,
  NetworkManager,
  LogoutManager,
  Controller,
  ExpiryStatus,
  actionTypes,
} from '@data-client/core';
export type {
  EndpointExtraOptions,
  FetchFunction,
  ResolveType,
  EndpointInterface,
  Schema,
  DenormalizeNullable,
  Denormalize,
  Normalize,
  NormalizeNullable,
  FetchAction,
  InvalidateAction,
  UnsubscribeAction,
  SubscribeAction,
  ResetAction,
  SetAction,
  ActionTypes,
  SetTypes,
  NetworkError,
  UnknownError,
  ErrorTypes,
  AbstractInstanceType,
  UpdateFunction,
  State,
  PK,
  Dispatch,
  Middleware,
  MiddlewareAPI,
  Manager,
  // used in Controller generic
  DataClientDispatch,
  GenericDispatch,
} from '@data-client/core';
export * from './components/index.js';
export * from './hooks/index.js';
export { StateContext, ControllerContext, StoreContext } from './context.js';
export type { Store } from './context.js';
export * as __INTERNAL__ from './internal.js';
export { usePromisifiedDispatch } from '@data-client/use-enhanced-reducer';
