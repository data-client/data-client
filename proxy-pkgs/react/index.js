export * from '@data-client/react';
