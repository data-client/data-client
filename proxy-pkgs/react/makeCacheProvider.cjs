module.exports = require('@data-client/react/makeCacheProvider');
