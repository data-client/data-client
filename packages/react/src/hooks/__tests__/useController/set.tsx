import { DataProvider } from '@data-client/react';
import { CoolerArticle } from '__tests__/new';
import nock from 'nock';

import { useQuery } from '../..';
import { makeRenderDataClient, act } from '../../../../../test';

export const payload = {
  id: 5,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};

export const createPayload = {
  id: 1,
  title: 'hi ho',
  content: 'whatever',
  tags: ['a', 'best', 'react'],
};
let renderDataClient: ReturnType<typeof makeRenderDataClient>;
let mynock: nock.Scope;

beforeEach(() => {
  renderDataClient = makeRenderDataClient(DataProvider);
  mynock = nock(/.*/).defaultReplyHeaders({
    'Access-Control-Allow-Origin': '*',
    'Content-Type': 'application/json',
  });
});
afterEach(() => {
  nock.cleanAll();
});

describe('set', () => {
  let errorspy: jest.MockInstance<typeof global.console.error, any>;
  beforeEach(() => {
    errorspy = jest
      .spyOn(global.console, 'error')
      .mockImplementation(() => {}) as any;
  });
  afterEach(() => {
    errorspy.mockRestore();
  });

  it('should update store when set is complete', async () => {
    const { result, controller } = renderDataClient(() => {
      return useQuery(CoolerArticle, { id: payload.id });
    });
    expect(result.current).toBeUndefined();
    let promise: any;
    act(() => {
      promise = controller.set(CoolerArticle, { id: 5 }, payload);
    });
    await act(() => promise);
    expect(result.current).toBeDefined();
    expect(result.current?.content).toEqual(payload.content);
    expect(result.current).toEqual(CoolerArticle.fromJS(payload));

    // type tests
    // TODO: move these to own unit tests if/when applicable
    () => {
      // @ts-expect-error
      controller.set(CoolerArticle, payload);
      controller.set(
        CoolerArticle,
        // @ts-expect-error
        payload.id,
        payload,
      );
    };
  });

  it('should update store with error', async () => {
    const { result, controller } = renderDataClient(() => {
      return useQuery(CoolerArticle, { id: payload.id });
    });
    expect(result.current).toBeUndefined();
    let promise: any;
    act(() => {
      promise = controller.set(CoolerArticle, { id: 5 }, 5);
    });
    expect(result.current).toBeUndefined();
    expect(errorspy.mock.calls.length).toBe(1);
    expect(errorspy.mock.calls).toMatchSnapshot();
  });
});
