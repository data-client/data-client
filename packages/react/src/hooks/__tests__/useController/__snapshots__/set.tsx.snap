// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set should update store with error 1`] = `
[
  [
    [Error: Unexpected input given to normalize. Expected type to be "object", found "number".

          Schema: {
  "name": "CoolerArticle",
  "schema": {
    "author": {
      "name": "User",
      "schema": {},
      "key": "User"
    }
  },
  "key": "CoolerArticle"
}
          Input: "5"],
  ],
]
`;
